cpp/src/lists/segmented_sort.cu

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/lists/sorting.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/copy.h>
+#include <cub/device/device_segmented_radix_sort.cuh>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf {
+namespace lists {
+namespace detail {
+
+struct SortPairs {
+  template <typename KeyT, typename ValueT, typename OffsetIteratorT>
+  void SortPairsAscending(KeyT const* keys_in,
+                          KeyT* keys_out,
+                          ValueT const* values_in,
+                          ValueT* values_out,
+                          int num_items,
+                          int num_segments,
+                          OffsetIteratorT begin_offsets,
+                          OffsetIteratorT end_offsets,
+                          rmm::cuda_stream_view stream)
+  {
+    rmm::device_buffer d_temp_storage;
+    size_t temp_storage_bytes = 0;
+    cub::DeviceSegmentedRadixSort::SortPairs(d_temp_storage.data(),
```

Comment:
Why can't we use `thrust::sort_by_key` instead?

Reply:
This is for single list column - faster than `thrust::sort_by_key`. Also with `DeviceSegmentedRadixSort`, there is room for improvement with multiple streams.

Reply:
Doesn't `thrust::sort_by_key` trivially map to `DeviceSegmentedRadixSort` when doing a numeric sort? Similar to how `thrust::sort` maps to `DeviceRadixSort`?

Reply:
It doesn't. `thrust::sort_by_key` is not a segmented sort. `thrust::sort_by_key` sorts `values` w.r.t to `key` sorted order.